manual/ahoy_config.md
# Ahoy configuration

## Prerequists
You have build your own hardware (or purchased one). The firmware is already loaded on the ESP and the WebUI is accessible from your browser.

## Start
But how do I get my data from the inverter?

The following steps are required:
1. Set the pinning to communicate with the radio module.
2. Check if Ahoy has a current time
3. Configure the inverter data (e.g. serialnumber)

### 1.) Set the pinning
Once you are in the web interface, you will find the "System Config" sub-item in the Setup area.

This is where you tell the ESP how you connected the radio module.
Note the schematics you saw earlier. - If you haven't noticed them yet, here's another table of connections.


#### OpenDTU Fusion (ESP32-S3)
| NRF24 Pin | ESP Pin|
|---------| --------|
| CS (4) | GPIO37
| CE (3)| GPIO38
| IRQ (8) | GPIO47
| SCLK (5)| GPIO36
| MOSI (6)| GPIO35
| MISO (7)| GPIO48

| CMT2300A  | Pin |
|---------| --------|
| CMT| Enabled |
| SCLK| GPIO6
| SDIO| GPIO5
| CSB| GPIO4
| FCSB| GPIO21
| GPIO3| GPIO8

### 2.) Set current time (standard: skip this step)
Ahoy needs a current date and time to talk to the inverter.
It works without, but it is recommended to include a time. This allows you to analyze information from the inverter in more detail.
Normally, a date/time should be automatically retrieved from the NTP server. However, it may happen that the firewall of some routers does not allow this.
In the section "Settings -> NTP Server" you can also get the time from your own computer. Or set up your own NTP server.

### 3.) Set inverter data

#### add new inverter
Now it's time to place the inverter. This is necessary because it is not the inverter that speaks first, but the DTU (Ahoy).

Each inverter has its own S.Nr. This also serves as an identity for communication between the DTU and the inverter.

The S.Nr is a 12-digit number. Check [here (german)](https://github.com/lumapu/ahoy/wiki/Hardware#wie-ist-die-serien-nummer-der-inverter-aufgebaut) for more information.

#### set pv-modules (not necessary)
Click on "Add Inverter" and enter the S.No. and a name. Please keep the name short!
![grafik](https://github.com/lumapu/ahoy/doc/screenshots/settings.png)

![grafik](https://github.com/lumapu/ahoy/doc/screenshots/inverterSettings.png)

In the upper tab "Inputs" you can enter the data of the solar modules. These are only used directly in Ahoy for calculation and have no influence on the inverter.

#### set radio parameter (not necessary, only for EU)
In the next tab "Radio" you can adjust the power and other parameters if necessary. However, these should be left as default (EU only).

#### advanced options (not necessary to be changed)
In the "Advanced" section, you can customize more settings.

Save and reboot.

## ✅ Done - Now check the live site


src/CHANGES.md
# Development Changes

## 0.8.49 - 2024-01-08
* fix send total values if inverter state is different from `OFF` #1331
* fix german language issues #1335

## 0.8.48 - 2024-01-07
* merge PR: pin selection for ESP-32 S2 #1334
* merge PR: enhancement: power graph display option #1330

## 0.8.47 - 2024-01-06
* reduce GxEPD2 lib to compile faster
* upgraded GxEPD2 lib to `1.5.3`
* updated espressif32 platform to `6.5.0`
* updated U8g2 to `2.35.9`
* started to convert deprecated functions of new ArduinoJson `7.0.0`
* started to have german translations of all variants (environments) #925 #1199
* merge PR: add defines for retry attempts #1329

## 0.8.46 - 2024-01-06
* improved communication

## 0.8.45 - 2024-01-05
* fix MqTT total values #1326
* start implementing a wizard for initial (WiFi) configuration #1199

## 0.8.44 - 2024-01-05
* fix MqTT transmission of data #1326
* live data is read much earlier / faster and more often #1272

## 0.8.43 - 2024-01-04
* fix display of sunrise in `/system` #1308
* fix overflow of `getLossRate` calculation #1318
* improved MqTT by marking sent data and improved `last_success` resends #1319
* added timestamp for `max ac power` as tooltip #1324 #1123 #1199
* repaired Power-limit acknowledge #1322
* fix `max_power` in `/visualization` was set to `0` after sunset

## 0.8.42 - 2024-01-02
* add LED to display whether it's night time or not. Can be reused as output to control battery system #1308
* merge PR: beautifiying typography, added spaces between value and unit for `/visualization` #1314
* merge PR: Prometheus add `getLossRate` and bugfixing #1315
* add loss rate to `/visualization` in the statistics window
* corrected `getLossRate` infos for MqTT and prometheus
* added information about working IRQ for NRF24 and CMT2300A to `/system`

## 0.8.41 - 2024-01-02
* fix display timeout (OLED) to 60s
* change offs to signed value

## 0.8.40 - 2024-01-02
* fix display of sunrise and sunset in `/system` #1308
* fix MqTT set power limit #1313

## 0.8.39 - 2024-01-01
* fix MqTT dis_night_comm in the morning #1309 #1286
* seperated offset for sunrise and sunset #1308
* powerlimit (active power control) now has one decimal place (MqTT / API) #1199
* merge Prometheus metrics fix #1310
* merge MI grid profile request #1306
* merge update documentation / readme #1305
* add `getLossRate` to radio statistics and to MqTT #1199

## 0.8.38 - 2023-12-31
* fix Grid-Profile JSON #1304

## 0.8.37 - 2023-12-30
* added grid profiles
* format version of grid profile 

# RELEASE 0.8.36 - 2023-12-30

## 0.8.35 - 2023-12-30
* added dim option for LEDS
* changed reload time for opendtufusion after update to 5s
* fix default interval and gap for communication
* fix serial number in exported json (was decimal, now correct as hexdecimal number)
* beautified factory reset
* added second stage for erase settings
* increased maximal number of inverters to 32 for opendtufusion board (ESP32-S3)
* fixed crash if CMT inverter is enabled, but CMT isn't configured

# RELEASE 0.8.34 - 2023-12-29

## 0.8.33 - 2023-12-29
* improved communication thx @rejoe2

## 0.8.32 - 2023-12-29
* fix `start` / `stop` / `restart` commands #1287
* added message, if profile was not read until now #1300
* added developer option to use 'syslog-server' instead of 'web-serail' #1299

## 0.8.31 - 2023-12-29
* added class to handle timeouts PR #1298

## 0.8.30 - 2023-12-28
* added info if grid profile was not found
* merge PR #1293
* merge PR #1295 fix ESP8266 pin settings
* merge PR #1297 fix layout for OLED displays

## 0.8.29 - 2023-12-27
* fix MqTT generic topic `comm_disabled` #1265 #1286
* potential fix of #1285 (reset yield day)
* fix fraction of yield correction #1280
* fix crash if `getLossRate` was read from inverter #1288 #1290
* reduce reload time for opendtufusion ethernet variant to 5 seconds
* added basic grid parser
* added ESP32-C3 mini environment #1289

## 0.8.28 - 2023-12-23
* fix bug heuristic
* add version information to clipboard once 'copy' was clicked
* add get loss rate @rejoe2
* improve communication @rejoe2

## 0.8.27 - 2023-12-18
* fix set power limit #1276

## 0.8.26 - 2023-12-17
* read grid profile as HEX (`live` -> click inverter name -> `show grid profile`)

## 0.8.25 - 2023-12-17
* RX channel ID starts with fixed value #1277
* fix static IP for Ethernet

## 0.8.24 - 2023-12-16
* fix NRF communication for opendtufusion ethernet variant

## 0.8.23 - 2023-12-14
* heuristics fix #1269 #1270
* moved `sendInterval` in settings, **important:** *will be reseted to 15s after update to this version*
* try to prevent access to radio classes if they are not activated
* fixed millis in serial log
* changed 'print whole trace' = `false` as default
* added communication loop duration in [ms] to serial console
* don't print Hex-Payload if 'print whole trace' == `false`

## 0.8.22 - 2023-12-13
* fix communication state-machine regarding zero export #1267

## 0.8.21 - 2023-12-12
* fix ethernet save inverter parameters #886
* fix ethernet OTA update #886
* improved radio statistics, fixed heuristic output for HMS and HMT inverters

## 0.8.20 - 2023-12-12
* improved HM communication #1259 #1249
* fix `loadDefaults` for ethernet builds #1263
* don't loop through radios which aren't in use #1264

## 0.8.19 - 2023-12-11
* added ms to serial log
* added (debug) option to configure gap between inverter requests

## 0.8.18 - 2023-12-10
* copied even more from the original heuristic code #1259
* added mDNS support #1262

## 0.8.17 - 2023-12-10
* possible fix of NRF with opendtufusion (without ETH)
* small fix in heuristics (if conditions made assignment not comparisson)

## 0.8.16 - 2023-12-09
* fix crash if NRF is not enabled
* updated heuristic #1080 #1259
* fix compile opendtufusion fusion ethernet

## 0.8.15 - 2023-12-09
* added support for opendtufusion fusion ethernet shield #886
* fixed range of HMS / HMT frequencies to 863 to 870 MHz #1238
* changed `yield effiency` per default to `1.0` #1243

src/config/config.h
//-----------------------------------------------------------------------------
// 2022 Ahoy, https://www.mikrocontroller.net/topic/525778
// Creative Commons - http://creativecommons.org/licenses/by-nc-sa/3.0/de/
//-----------------------------------------------------------------------------

#ifndef __CONFIG_H__
#define __CONFIG_H__


// globally used
#define DEF_PIN_OFF         255


//-------------------------------------
// WIFI CONFIGURATION
//-------------------------------------

// Fallback WiFi Info
#define FB_WIFI_SSID    "YOUR_WIFI_SSID"
#define FB_WIFI_PWD     "YOUR_WIFI_PWD"


// Access Point Info
// In case there is no WiFi Network or Ahoy can not connect to it, it will act as an Access Point

#define WIFI_AP_SSID    "AHOY-DTU"
#define WIFI_AP_PWD     "esp_8266"

// If the next line is uncommented, Ahoy will stay in access point mode all the time
//#define AP_ONLY

// timeout for automatic logoff (20 minutes)
#define LOGOUT_TIMEOUT      (20 * 60)

//-------------------------------------
// CONFIGURATION - COMPILE TIME
//-------------------------------------

// ethernet

#if defined(ETHERNET)
    #define ETH_SPI_HOST            SPI2_HOST
    #define ETH_SPI_CLOCK_MHZ       25
    #ifndef DEF_ETH_IRQ_PIN
        #define DEF_ETH_IRQ_PIN     4
    #endif
    #ifndef DEF_ETH_MISO_PIN
        #define DEF_ETH_MISO_PIN    12
    #endif
    #ifndef DEF_ETH_MOSI_PIN
        #define DEF_ETH_MOSI_PIN    13
    #endif
    #ifndef DEF_ETH_SCK_PIN
        #define DEF_ETH_SCK_PIN     14
    #endif
    #ifndef DEF_ETH_CS_PIN
        #define DEF_ETH_CS_PIN      15
    #endif
#else /* defined(ETHERNET) */
// time in seconds how long the station info (ssid + pwd) will be tried
#define WIFI_TRY_CONNECT_TIME   30

// time during the ESP will act as access point on connection failure (to
// station) in seconds
#define WIFI_AP_ACTIVE_TIME     60
#endif /* defined(ETHERNET) */

// default device name
#define DEF_DEVICE_NAME         "AHOY-DTU"

// default pinout (GPIO Number)
#if defined(ESP32)
    // this is the default ESP32 (son-S) pinout on the WROOM modules for VSPI,
    // for the ESP32-S3 there is no sane 'default', as it has full flexibility
    // to map its two HW SPIs anywhere and PCBs differ materially,
    // so it has to be selected in the Web UI
    #ifndef DEF_NRF_CS_PIN
        #define DEF_NRF_CS_PIN          5
    #endif
    #ifndef DEF_NRF_CE_PIN
        #define DEF_NRF_CE_PIN          4
    #endif
    #ifndef DEF_NRF_IRQ_PIN
        #define DEF_NRF_IRQ_PIN         16
    #endif
    #ifndef DEF_NRF_MISO_PIN
        #define DEF_NRF_MISO_PIN        19
    #endif
    #ifndef DEF_NRF_MOSI_PIN
        #define DEF_NRF_MOSI_PIN        23
    #endif
    #ifndef DEF_NRF_SCLK_PIN
        #define DEF_NRF_SCLK_PIN        18
    #endif

    #ifndef DEF_CMT_SCLK
        #define DEF_CMT_SCLK            12
    #endif
    #ifndef DEF_CMT_SDIO
        #define DEF_CMT_SDIO            14
    #endif
    #ifndef DEF_CMT_CSB
        #define DEF_CMT_CSB             27
    #endif
    #ifndef DEF_CMT_FCSB
        #define DEF_CMT_FCSB            26
    #endif
    #ifndef DEF_CMT_IRQ
        #define DEF_CMT_IRQ             34
    #endif
    #ifndef DEF_MOTION_SENSOR_PIN
        #define DEF_MOTION_SENSOR_PIN   DEF_PIN_OFF
    #endif
#else
    #ifndef DEF_NRF_CS_PIN
        #define DEF_NRF_CS_PIN          15
    #endif
    #ifndef DEF_NRF_CE_PIN
        #define DEF_NRF_CE_PIN          0
    #endif
    #ifndef DEF_NRF_IRQ_PIN
        #define DEF_NRF_IRQ_PIN         2
    #endif
    // these are given to relay the correct values via API
    // they cannot actually be moved for ESP82xx models
    #ifndef DEF_NRF_MISO_PIN
        #define DEF_NRF_MISO_PIN        12
    #endif
    #ifndef DEF_NRF_MOSI_PIN
        #define DEF_NRF_MOSI_PIN        13
    #endif
    #ifndef DEF_NRF_SCLK_PIN
        #define DEF_NRF_SCLK_PIN        14
    #endif
    #ifndef DEF_MOTION_SENSOR_PIN
        #define DEF_MOTION_SENSOR_PIN   A0
    #endif
#endif
#ifndef DEF_LED0
    #define DEF_LED0                DEF_PIN_OFF
#endif
#ifndef DEF_LED1
    #define DEF_LED1                DEF_PIN_OFF
#endif
#ifdef LED_ACTIVE_HIGH
    #define LED_HIGH_ACTIVE         true
#else
    #define LED_HIGH_ACTIVE         false
#endif

// number of packets hold in buffer
#define PACKET_BUFFER_SIZE      30

// number of configurable inverters
#if defined(ESP32)
    #if defined(CONFIG_IDF_TARGET_ESP32S3)
        #define MAX_NUM_INVERTERS   32
    #else
        #define MAX_NUM_INVERTERS   16
    #endif
#else
    #define MAX_NUM_INVERTERS   4
#endif

// default send interval
#define SEND_INTERVAL           15

// maximum human readable inverter name length
#define MAX_NAME_LENGTH         16

// maximum buffer length of packet received / sent to RF24 module
#define MAX_RF_PAYLOAD_SIZE     32

// maximum total payload buffers (must be greater than the number of received frame fragments)
#define MAX_PAYLOAD_ENTRIES     20

// number of seconds since last successful response, before inverter is marked inactive
#define INVERTER_INACT_THRES_SEC    5*60

// number of seconds since last successful response, before inverter is marked offline
#define INVERTER_OFF_THRES_SEC      15*60

// threshold of minimum power on which the inverter is marked as inactive
#define INACT_PWR_THRESH        3

// Timezone
#define TIMEZONE                1

// default NTP server uri
#define DEF_NTP_SERVER_NAME     "pool.ntp.org"

// default NTP server port
#define DEF_NTP_PORT            123

// NTP refresh interval in ms (default 12h)
#define NTP_REFRESH_INTERVAL    12 * 3600 * 1000

// default mqtt interval
#define MQTT_INTERVAL           90

// default MQTT broker uri
#define DEF_MQTT_BROKER         "\0"

// default MQTT port
#define DEF_MQTT_PORT           1883

// default MQTT user
#define DEF_MQTT_USER           "\0"

// default MQTT pwd
#define DEF_MQTT_PWD           "\0"

// default MQTT topic
#define DEF_MQTT_TOPIC         "inverter"

// discovery prefix
#define MQTT_DISCOVERY_PREFIX   "homeassistant"

// reconnect delay
#define MQTT_RECONNECT_DELAY    5000

#if __has_include("config_override.h")
    #include "config_override.h"
#endif

#endif /*__CONFIG_H__*/
